src/brod_group_subscriber.erl

```diff
@@ -389,55 +357,51 @@ terminate(_Reason, #state{}) ->
-                            , cb_state         = CbState
-                            } = State) ->
-  %% process new message only when there is no pending ack
-  {AckNow, NewCbState} = F(CbState),
+handle_messages(_Topic, _Partition, [], State) ->
+  State;
+handle_messages(Topic, Partition, [Msg | Rest], State) ->
+  #kafka_message{offset = Offset} = Msg,
+  #state{cb_module = CbModule, cb_state = CbState} = State,
+  AckRef = {Topic, Partition, Offset},
+  {AckNow, NewCbState} =
+    case CbModule:handle_message(Topic, Partition, Msg, CbState) of
+      {ok, NewCbState_} ->
+        {true, NewCbState_};
+      {ok, ack, NewCbState_} ->
+        {false, NewCbState_};
+      Unknown ->
+        erlang:error({bad_return_value,
+                     {CbModule, handle_message, Unknown}})
+    end,
+  State1 = State#state{cb_state = NewCbState},
   NewState =
-    State#state{ pending_ack      = AckRef
-               , pending_messages = Rest
-               , cb_state         = NewCbState
-               },
-  case AckNow of
-    true  -> handle_ack(AckRef, NewState);
-    false -> {ok, NewState}
-  end;
-maybe_process_message(State) ->
-  {ok, State}.
+    case AckNow of
+      true  -> handle_ack(AckRef, State1);
+      false -> State1
+    end,
+  handle_messages(Topic, Partition, Rest, NewState).
 
--spec handle_ack(ack_ref(), #state{}) -> {ok, #state{}}.
-handle_ack(AckRef, #state{ pending_ack      = AckRef
-                         , pending_messages = Messages
-                         , generationId     = GenerationId
-                         , consumers        = Consumers
-                         , coordinator      = Coordinator
-                         } = State0) ->
+-spec handle_ack(ack_ref(), #state{}) -> #state{}.
+handle_ack(AckRef, #state{ generationId = GenerationId
+                         , consumers    = Consumers
+                         , coordinator  = Coordinator
+                         } = State) ->
   {Topic, Partition, Offset} = AckRef,
-  State1 =
-    case lists:keyfind({Topic, Partition},
-                       #consumer.topic_partition, Consumers) of
-      #consumer{consumer_pid = ConsumerPid} = Consumer ->
-        ok = brod:consume_ack(ConsumerPid, Offset),
-        ok = brod_group_coordinator:ack(Coordinator, GenerationId,
-                                        Topic, Partition, Offset),
-        NewConsumer = Consumer#consumer{acked_offset = Offset},
-        NewConsumers = lists:keyreplace({Topic, Partition},
-                                        #consumer.topic_partition,
-                                        Consumers, NewConsumer),
-      State0#state{consumers = NewConsumers};
+  case lists:keyfind({Topic, Partition},
+                     #consumer.topic_partition, Consumers) of
+    #consumer{consumer_pid = ConsumerPid} = Consumer ->
+      ok = brod:consume_ack(ConsumerPid, Offset),
+      ok = brod_group_coordinator:ack(Coordinator, GenerationId,
```

Comment:
NO, for 2 reasons:
1. We do not know the cost of each `CbModule:handle_message/2`, in case it's slow, acking only the last one will delay the pre-fetches.
2. `brod:consume_ack/2` and `brod_group_coordinator:ack/5` are quite cheap.